cpp/include/raft/spatial/knn/brute_force.cuh

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "detail/knn_brute_force_faiss.cuh"
+#include "detail/selection_faiss.cuh"
+#include <raft/core/device_mdspan.hpp>
+
+namespace raft::spatial::knn {
+
+/**
+ * @brief Performs a k-select across row partitioned index/distance
+ * matrices formatted like the following:
+ * row1: k0, k1, k2
+ * row2: k0, k1, k2
+ * row3: k0, k1, k2
+ * row1: k0, k1, k2
+ * row2: k0, k1, k2
+ * row3: k0, k1, k2
+ *
+ * etc...
+ *
+ * @tparam idx_t
+ * @tparam value_t
+ * @param[in] handle
+ * @param[in] in_keys matrix of input keys (size n_samples * n_parts * k)
+ * @param[in] in_values matrix of input values (size n_samples * n_parts * k)
+ * @param[out] out_keys matrix of output keys (size n_samples * k)
+ * @param[out] out_values matrix of output values (size n_samples * k)
+ * @param[in] n_samples number of rows in each part
+ * @param[in] k number of neighbors for each part
+ * @param[in] translations optional vector of starting index mappings for each partition
+ */
+template <typename idx_t = int64_t, typename value_t = float>
```

Comment:
Per @mhoemmen I also removed defaults for mdspan-related templates that are inferred from my `raft::linalg` PR. I think we can apply this across the whole PR